src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,178 @@
+    protected def getPartitions: Array[Partition] =
+      parts
+
+    /** Optionally overridden by subclasses to specify how they are partitioned. */
+    @transient override val partitioner: Option[Partitioner] =
+      Some(new GridPartitioner(rowBlocks, colBlocks, 1, 1))
+  }
+
+  case class IntPartition(index: Int) extends Partition { }
+
+  private def gridPartition(m: BlockMatrix): BlockMatrix = {
+    val p = new GridPartitioner(m.numRowBlocks, m.numColBlocks, 1, 1)
+    new BlockMatrix(m.blocks.partitionBy(p), m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols())
+  }
+
+  private def ensureGridPartitioning(m: BlockMatrix): BlockMatrix = m.blocks.partitioner match {
+    case Some(gp: GridPartitioner) if (gp.rowsPerPart == 1 && gp.colsPerPart == 1) =>
+      m
+    case Some(gp: GridPartitioner) =>
+      logDebug(s"Repartitioning a matrix (slow), $m, with a grid partitioner that didn't have 1 block per partition, had: ${gp.rowsPerPart} x ${gp.colsPerPart}")
+      gridPartition(m)
+    case Some(p) =>
+      logDebug(s"Repartitioning a matrix (slow), $m, with a non-grid partitioner: $p")
+      gridPartition(m)
+    case None =>
+      logDebug(s"Partitioning a matrix (slow), $m, with a no partitioner")
+      gridPartition(m)
+  }
+
+  def multiply(preL: BlockMatrix, preR: BlockMatrix): BlockMatrix = {
+    require(preL.numCols() == preR.numRows(),
+      s"""The number of columns of the left matrix and the number of rows of the right
+          matrix must be equal: ${preL.numRows()} x ${preL.numCols()}, ${preR.numRows()}
+          x ${preR.numCols()}. If you think they should be equal, try setting the
+          dimensions of A and B explicitly while initializing them.""".stripMargin)
+    // FIXME: don't require same blocksize on each matrix
+    require(preL.colsPerBlock == preR.rowsPerBlock,
+      s"""The number of columns in blocks of the left matrix and the number of rows in
+          blocks of the right matrix must be equal: ${preL.rowsPerBlock} x
+          ${preL.colsPerBlock}, ${preR.rowsPerBlock} x ${preR.colsPerBlock}. Generally,
+          all matrices should use square blocks of the same dimension.""".stripMargin)
+
+    val l = ensureGridPartitioning(preL)
+    val r = ensureGridPartitioning(preR)
+    new BlockMatrix(new BlockMatrixMultiplyRDD(l, r), l.rowsPerBlock, r.colsPerBlock, l.numRows(), r.numCols())
+  }
+
+  class BlockMatrixTransposeRDD(m: BlockMatrix)
+      extends RDD[((Int, Int), Matrix)](m.blocks.sparkContext, Seq[Dependency[_]](new OneToOneDependency(m.blocks))) {
+
+    def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), Matrix)] = {
+      val it = m.blocks.compute(split, context).map { case ((i, j), m) => ((j, i), m.transpose) }
+      val r = it.next()
+      assert(!it.hasNext, s"Expected iterator of size one, but got blocks: ${r +: it.toSeq}, from $m with partitioner ${m.blocks.partitioner}. ${m.blocks.toDebugString}")
```

Comment:
`ensureGridPartitioning` ensures that it has a grid partitioner that claims to have 1 block per partition. I guess I'm just paranoid because the type is `Iterator` and the call to `ensureGridPartitioner` is physically distant in this file.

I'll remove it and make this class private.